### PR TITLE
doc: update the ACRN on QEMU tutorial to v2.5

### PR DESCRIPTION
**Pull request by gvancuts**
Update the "Running ACRN on QEMU" tutorial to be based on ACRN
v2.4 and Ubuntu 20.04.

Signed-off-by: Geoffroy Van Cutsem <geoffroy.vancutsem@intel.com>

### CODE REVIEW

File: doc/tutorials/acrn_on_qemu.rst [hunk 1]
```diff
@@ -64,35 +67,42 @@ Prepare Service VM (L1 Guest)
...
          :align: center
 
-3. To login to the Service VM guest, find the IP address of the guest to SSH. This can be done via the
-   virsh command as shown below,
+   c. The Service VM (guest) will be restarted once the installation is complete.
 
-   .. figure:: images/acrn_qemu_3.png
-      :align: center
+#. Login to the Service VM guest. Find the IP address of the guest and use it to connect
+   via SSH. The IP address can be retrieved using via the ``virsh`` command as shown below.
```
Comment (dbkinder) [inline doc/tutorials/acrn_on_qemu.rst:85]:
```suggestion
   via SSH. The IP address can be retrieved using the ``virsh`` command as shown below.
```

File: doc/tutorials/acrn_on_qemu.rst [hunk 2]
```diff
@@ -64,35 +67,42 @@ Prepare Service VM (L1 Guest)
...
+
+   .. code-block:: console
 
-4. Once ACRN hypervisor is enabled, the above virsh command might not display the IP. So enable Serial console by,
+      $ virsh domifaddr ACRNSOS
+       Name       MAC address          Protocol     Address
+      -------------------------------------------------------------------------------
+       vnet0      52:54:00:72:4e:71    ipv4         192.168.122.31/24
+
+#. Once logged into the Service VM, enable the serial console as, once ACRN is enabled,
```
Comment (dbkinder) [inline doc/tutorials/acrn_on_qemu.rst:94]:
```suggestion
#. Once logged into the Service VM, enable the serial console. Once ACRN is enabled,
```

File: doc/tutorials/acrn_on_qemu.rst [hunk 3]
```diff
@@ -170,107 +199,66 @@ Install ACRN Hypervisor
...
+      [    0.000000] Hypervisor detected: ACRN
+      [    2.337176] ACRNTrace: Initialized acrn trace module with 4 cpu
+      [    2.368358] ACRN HVLog: Initialized hvlog module with 4 cpu
+      [    2.727905] systemd[1]: Set hostname to <ACRNSOS>.
 
-   a. Copy files ``misc/acrnbridge/acrn.network`` and ``misc/acrnbridge/tap0.netdev`` from the cloned ACRN repo to :file:`/usr/lib/system/network`.
-   b. Rename ``acrn.network`` to ``50-acrn.network``.
-   c. Rename ``tap0.netdev`` to ``50-tap0.netdev``.
+   .. note::
+      When shutting down the Service VM, make sure to cleanly destroy it
```
Comment (dbkinder) [inline doc/tutorials/acrn_on_qemu.rst:228]:
```suggestion
      When shutting down the Service VM, make sure to cleanly destroy it with these commands,
```

File: doc/tutorials/acrn_on_qemu.rst [hunk 4]
```diff
@@ -170,107 +199,66 @@ Install ACRN Hypervisor
...
 
-      $ git clone https://github.com/projectacrn/acrn-kernel
-      $ cd acrn-kernel
+      $ cd ~/acrn-kernel
       $ cp kernel_config_uos .config
+      $ make olddefconfig
       $ make
 
-   Once the User VM kernel is built successfully, copy ``arch/x86/boot/bzImage`` to  ACRNSOS (L1 guest) and rename this to ``bzImage_uos``. Need this to launch the User VM (L2 guest)
+#. Copy the User VM kernel to your home folder, we will use it to launch the User VM (L2 guest)
```
Comment (dbkinder) [inline doc/tutorials/acrn_on_qemu.rst:247]:
```suggestion
#. Copy the User VM kernel to your home folder. We will use it to launch the User VM (L2 guest)
```

**Review by dbkinder — CHANGES_REQUESTED** [review]
thanks @gvancuts

**Comment by gvancuts** [comment]
Thanks @dbkinder , I have merged all the changes you proposed.

**Comment by dbkinder** [comment]
ok to verify

**Comment by dbkinder** [comment]
@NanlinXie @wenlingz Would you figure out why this doc-change only PR is failing build tests...

**Comment by wenlingz** [comment]
> @NanlinXie @wenlingz Would you figure out why this doc-change only PR is failing build tests...

done

**Comment by wenlingz** [comment]
ok to verify

**Review by dbkinder — APPROVED** [review]
LGTM, but do we now want to update this for v2.5?

**Comment by dbkinder** [comment]
@gvancuts is this ready to not be a "draft" so we can merge and publish?  Do you want to update to work with v2.5 since that's the upcoming release?

**Comment by gvancuts** [comment]
> @gvancuts is this ready to not be a "draft" so we can merge and publish? Do you want to update to work with v2.5 since that's the upcoming release?

I'm still keeping this as a draft because I am not able to run ACRN on QEMU using this tutorial. @fuzhongl is able to make it work but we do not understand yet what the difference is between his environment and mine that makes it work for him but fail for me.

Once that's figured out, I'll mark it as "Ready for review" and hopefully we can merge this. I may update this to v2.5 too if I have time... we'll see.

**Comment by gvancuts** [comment]
I cannot test on v2.5 for now because of #6102 (compilation failure).

**Comment by gvancuts** [comment]
It's now running on my side so moving this from draft to "ready for review"!

**Comment by dbkinder** [comment]
> Thanks @dbkinder , I have merged all the changes you proposed.

It doesn't look like the proposed changes were made?

**Review by dbkinder — CHANGES_REQUESTED** [review]
@gvancuts it doesn't look like the suggested edits were actually made?

**Comment by fuzhongl** [comment]
> It's now running on my side so moving this from draft to "ready for review"!

@gvancuts 
Sometimes ACRN will assert with following error log:
`Assertion failed in file arch/x86/vtd.c,line 256 : fatal error`
Dev team is trying to narrow down and root cause it.

This issue can be workaround with below command in native:
`sudo systemctl set-default multi-user.target`

Could you please help to add this information in this PR before the final fix?
Thanks!

**Comment by gvancuts** [comment]
> @gvancuts it doesn't look like the suggested edits were actually made?

My bad, not sure what I did (or did not do as the case may be). I have now committed all these changes.

**Comment by gvancuts** [comment]
> @gvancuts
> Sometimes ACRN will assert with following error log:
> `Assertion failed in file arch/x86/vtd.c,line 256 : fatal error`

I have seen this too, on rather few occasions thankfully, but it would be good to root-cause it!

> Dev team is trying to narrow down and root cause it.

Great - thanks!
 
> This issue can be workaround with below command in native:
> `sudo systemctl set-default multi-user.target`

Hmm... I believe that error is happening when the kernel is booting up, how confident are you that it's something later when the user-space is coming up that's causing this?
 
> Could you please help to add this information in this PR before the final fix?

Are we confident this is really a workaround?

**Comment by fuzhongl** [comment]
> > @gvancuts
> > Sometimes ACRN will assert with following error log:
> > `Assertion failed in file arch/x86/vtd.c,line 256 : fatal error`
> 
> I have seen this too, on rather few occasions thankfully, but it would be good to root-cause it!
> 
> > Dev team is trying to narrow down and root cause it.
> 
> Great - thanks!
> 
> > This issue can be workaround with below command in native:
> > `sudo systemctl set-default multi-user.target`
> 
> Hmm... I believe that error is happening when the kernel is booting up, how confident are you that it's something later when the user-space is coming up that's causing this?
> 
> > Could you please help to add this information in this PR before the final fix?
> 
> Are we confident this is really a workaround?

@gvancuts Developer think this issue is related with GPU who is using DMA at that moment.
With `sudo systemctl set-default multi-user.target` , the issue can't be reproduced for more than tens of qemu `destroy` and `start`. Anyway, Dev team will root cause and fix it later.

**Comment by gvancuts** [comment]
@fuzhongl @dbkinder PTAL

**Comment by fuzhongl** [comment]
> @fuzhongl @dbkinder PTAL

@gvancuts Thanks for update!
There is another issue for v2.4 as mentioned in 5938:
```
detect_hardware_support, physical-address width (40) over maximum physical-address width (39)
[0us][cpu=0][(null)][sev=1][seq=2]:PANIC: init_pcpu_pre line: 132
[0us][cpu=0][(null)][sev=1][seq=3]:hardware not support!
```

Need following change:
```
diff --git a/hypervisor/include/arch/x86/page.h b/hypervisor/include/arch/x86/page.h
index 1094fa30d..e7c042903 100644
--- a/hypervisor/include/arch/x86/page.h
+++ b/hypervisor/include/arch/x86/page.h
@@ -14,7 +14,7 @@
 #define PAGE_SIZE      (1U << PAGE_SHIFT)
 #define PAGE_MASK      0xFFFFFFFFFFFFF000UL

-#define MAXIMUM_PA_WIDTH       39U     /* maximum physical-address width */
+#define MAXIMUM_PA_WIDTH       40U     /* maximum physical-address width */

 /* size of the low MMIO address space: 2GB */
 #define PLATFORM_LO_MMIO_SIZE  0x80000000UL
```
Will merge it to 2.4 branch or highlight this in this PR?
@gvancuts What's your prefer?

**Comment by gvancuts** [comment]
> I cannot test on v2.5 for now because of #6102 (compilation failure).

#6102 has been resolved and the fix backported to the `release_2.5` branch. As a result of this, I have updated the tutorial to be based on ACRN v2.5.

**Important note**: this tutorial uses the `v2.5` tag for **both** the `acrn-hypervisor` and `acrn-kernel` repositories. These tags do not exist yet and my own local testing was based on the `release_2.5` branches.

**Comment by gvancuts** [comment]
> Will merge it to 2.4 branch or highlight this in this PR?
> @gvancuts What's your prefer?

I have updated the PR to use v2.5 as the new baseline now that the blocking issue has been resolved. I don't think we should merge this to our 2.4 release. We _could_ add a note to show the patch in case someone wants to use the v2.4 release but I think at this stage it's actually better to redirect everyone to v2.5 so I wouldn't even include that note.

**Review by NanlinXie — APPROVED** [review]

**Review by dbkinder — APPROVED** [review]
LGTM, thanks!

**Pull request merged**